gdrapi.c

```diff
@@ -335,6 +344,90 @@ static void gdr_init_cpu_flags()
 
 // note: more than one implementation may be compiled in
 
+static void unroll8_memcpy(void *bar_ptr, const void *h_ptr, size_t size)
+{
+    const uint64_t *r = (const uint64_t *)h_ptr;
+    uint64_t *w = (uint64_t *)bar_ptr;
+    size_t nw = size / sizeof(*r);
+    assert(size % sizeof(*r) == 0);
+
+    while (nw) {
+        if (0 == (nw & 3)) {
+            uint64_t r0 = r[0];
+            uint64_t r1 = r[1];
+            uint64_t r2 = r[2];
+            uint64_t r3 = r[3];
+            w[0] = r0;
+            w[1] = r1;
+            w[2] = r2;
+            w[3] = r3;
+            r += 4;
+            w += 4;
+            nw -= 4;
+        } else if (0 == (nw & 1)) {
+            uint64_t r0 = r[0];
+            uint64_t r1 = r[1];
+            w[0] = r0;
+            w[1] = r1;
+            r += 2;
+            w += 2;
+            nw -= 2;
+        } else {
+            w[0] = r[0];
+            ++w;
+            ++r;
+            --nw;
+        }
+    }
+}
+
+static void unroll4_memcpy(void *bar_ptr, const void *h_ptr, size_t size)
```

Comment:
Change the name "bar_ptr" and "h_ptr" because it is incorrect when this function is called from "gdr_copy_from_bar".